### PR TITLE
Int2019 33 transfer funds

### PR DESCRIPTION
**Pull request by radulambrino**
Transfer Funds + RequestTransfer(sender view)

### CODE REVIEW

File: ClientApp/viking-vault/src/Common/Toggle.css [hunk 1]
```diff
@@ -0,0 +1,68 @@
+.switch {
```
Comment (vasi-axinte) [inline ClientApp/viking-vault/src/Common/Toggle.css]: This is not suggestive enough and it should be changed to something more meaningful.

**Review by vasi-axinte — CHANGES_REQUESTED** [review]

File: ClientApp/viking-vault/src/TransferFunds/TransferFundsPage.tsx [hunk 2]
```diff
@@ -0,0 +1,328 @@
...
+    }
+
+    private setCurrency = (e: any) => {
+      this.setState({
+        currency: e.target.value
+      });
+      this.getTotalBalance();
+    }
+
+    handleChangerequestTransferState = () =>
```
Comment (vasi-axinte) [inline ClientApp/viking-vault/src/TransferFunds/TransferFundsPage.tsx]: ChangeRequest are two different words, so they should start with capital letter.

File: ClientApp/viking-vault/src/TransferFunds/TransferFundsPage.tsx [hunk 3]
```diff
@@ -0,0 +1,328 @@
...
+                method: "GET",
+                headers: {
+                  'Accept': 'application/json',
+                  'Content-Type': 'application/json',
+                  'x-access-token': token.toString()
+                }
+            })
+            .then(response => response.json())
+            .then(result => {
+                if(this.state.currency === "RON") {
```
Comment (vasi-axinte) [inline ClientApp/viking-vault/src/TransferFunds/TransferFundsPage.tsx]: This method became too long. Should be refactored to be smaller. One can notice that all those if`s differ slightly, therefore they can be easily extracted to a method.

File: ClientApp/viking-vault/src/TransferFunds/TransferFundsPage.tsx [hunk 4]
```diff
@@ -0,0 +1,328 @@
...
+                        totalBalance: result[3].balance.toFixed(2)
+                    });
+
+                }
+            });
+        }
+    }
+
+    transferMoney = () => {
+        var data = {
```
Comment (vasi-axinte) [inline ClientApp/viking-vault/src/TransferFunds/TransferFundsPage.tsx]: Same here. Method too long. In order to understand it, I need to read a lot of code. That`s a lot of effort.

File: ClientApp/viking-vault/src/TransferFunds/TransferFundsPage.tsx [hunk 5]
```diff
@@ -0,0 +1,328 @@
...
+                  }
+                  this.handleIsButtonDisabled(); 
+              });
+          }
+        }
+    }
+
+    requestTransfer = () =>
+    {
+      var data = {
```
Comment (vasi-axinte) [inline ClientApp/viking-vault/src/TransferFunds/TransferFundsPage.tsx]: Same here.

File: Server/VikingVault.API/Controllers/TransferRequestsController.cs [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+                        return Ok("Succesfully requested " + transferRequestData.Amount + "!");
+                    }
+                    else
+                    {
+                        return Ok("Select a different user to request money from!");
+                    }                   
+                }
+                else
+                {
+                    return NotFound("Request to server unsuccesful.");
```
Comment (vasi-axinte) [inline Server/VikingVault.API/Controllers/TransferRequestsController.cs]: The request was ok, the problem would be that it didn`t find something that was requested.

File: Server/VikingVault.Services/Abstractions/IBankAccountService.cs [hunk 7]
```diff
@@ -11,5 +11,7 @@ public interface IBankAccountService
         List<BankAccount> GetBankAccounts(string token);
         void UpdateBankAccount();
         BankAccount ChangeBalance(UpdateBankAccountModel updatedBankAccount);
+        BankAccount AddMoneyToAccount(User user, string currency, float amountToBeAdded);
```
Comment (vasi-axinte) [inline Server/VikingVault.Services/Abstractions/IBankAccountService.cs]: How comes that you named a method "AddMoneyToAccount" but no account was sent as parameter?

File: Server/VikingVault.Services/TransferFundsService.cs [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+            _dbContext = dbContext;
+            _bankAccountService = bankAccountService;
+            _transactionService = transactionService;
+            _userCardService = userCardService;
+            _userService = userService;
+        }
+
+        public bool? TransferFunds(TransferFundsModel transferData)
+        {
+            int idSender = transferData.Sender.Id;
```
Comment (vasi-axinte) [inline Server/VikingVault.Services/TransferFundsService.cs:34]: You have so many if`s in this method... You shall refactor it in order to use fewer if`s or at least, extract some of them to separate methods.

**Review by vasi-axinte — CHANGES_REQUESTED** [review]

File: Server/VikingVault.API/Controllers/TransferRequestsController.cs (hunk 6 above)
Comment (radulambrino) [reply Server/VikingVault.API/Controllers/TransferRequestsController.cs]: Fair enough. I was unsure about this kind of situations and should've asked you when you was here.

File: Server/VikingVault.Services/Abstractions/IBankAccountService.cs (hunk 7 above)
Comment (radulambrino) [reply Server/VikingVault.Services/Abstractions/IBankAccountService.cs]: In this case, I am using the method in the controller, so I wanted to pass the responsibility of creating the account model to the method. I thought of adding only one BankAccount parameter to the method, but in the end I chose to do it this way beacause it reduces the number of lines in the top level code. In this case do you suggest changing the method name or the signature?

File: ClientApp/viking-vault/src/Common/Toggle.tsx [hunk 9]
```diff
@@ -0,0 +1,19 @@
+import * as React from "react";
+import './Toggle.css';
+import { constants } from "../Resources/Constants";
+
+interface IToggleState{
+    toggleSwitch: any
+}
+
+function Toggle (props: IToggleState)
+{
```
Comment (Slitthe) [inline ClientApp/viking-vault/src/Common/Toggle.tsx]: JavaScript convention is that the first code block is on the same line
blah {
}

File: ClientApp/viking-vault/src/TransferFunds/TransferFundsPage.tsx [hunk 10]
```diff
@@ -0,0 +1,328 @@
...
+
+    handleChangedCardNumber = (e : any) =>
+    {
+        this.setState({
+            cardNumber: e.target.value
+        })
+    }
+
+    handleChangedTransferDetails = (e : any) =>
+    {
```
Comment (Slitthe) [inline ClientApp/viking-vault/src/TransferFunds/TransferFundsPage.tsx]: Same here and above and below (code block positioning in JavaScript)

File: ClientApp/viking-vault/src/TransferFunds/TransferFundsPage.tsx [hunk 11]
```diff
@@ -0,0 +1,328 @@
...
+                    this.setState({
+                        totalBalance: result[1].balance.toFixed(2)
+                    });
+                }
+                if(this.state.currency === "USD") {
+                    this.setState({
+                        totalBalance: result[2].balance.toFixed(2)
+                    });
+                }
+                if(this.state.currency === "YEN") {
```
Comment (Slitthe) [inline ClientApp/viking-vault/src/TransferFunds/TransferFundsPage.tsx]: Please find a way to get rid of these multiple if's, they're really a bad practice as if they are hard to modify and it's also a lot of repeating code.

Proposed solution
-- create a dictionary for currencies and their indexes (or indices, whatever it's spelled) in the result array
-- return something like this `result[currenyMap[this.state.currency]].balance.toFixed(2)`
(of course also handle the default value, you can use Object.HasOwnProperty("propName") too see if an object has a property with that name)

File: Server/VikingVault.API/Controllers/TransferFundsController.cs [hunk 12]
```diff
@@ -0,0 +1,59 @@
...
+                {
+                    User sender = _userService.GetUserFromToken(token);
+                    TransferFundsModel transferFundsData = transferFundsDTO.convertDTOtoTransferFundsModel(sender);
+
+                    _transferFundsService.TransferFunds(transferFundsData);
+                    return Ok("Succesfully transfered " + transferFundsData.AmountSent + "!");
+                }
+                else
+                {
+                    return NotFound("Request to server unsuccesful.");
```
Comment (Slitthe) [inline Server/VikingVault.API/Controllers/TransferFundsController.cs]: Could you be more explicit with this error message?

File: Server/VikingVault.API/Controllers/TransferRequestsController.cs [hunk 13]
```diff
@@ -0,0 +1,68 @@
...
+                }
+            }
+            catch (Exception e)
+            {
+                if (e is NoCardAttachedToUserException || e is TransactionException || e is DatabaseException)
+                {
+                    return NotFound(e.Message);
+                }
+
+                return NotFound("Unknown error.");
```
Comment (Slitthe) [inline Server/VikingVault.API/Controllers/TransferRequestsController.cs:64]: Usually if something goes wrong but you don't know what's wrong it's an 500 Internal Server Error, a NotFound has the status code of 404 which simply menans that the resource that you've asked couldn't be found, but everything else worked as expected. But in here the NotFound is thrown at the end of the `catch( Exception)`

File: Server/VikingVault.Services/BankAccountService.cs [hunk 14]
```diff
@@ -55,6 +55,35 @@ public BankAccount ChangeBalance(UpdateBankAccountModel updatedBankAccount)
...
+            var updatedAccount = new UpdateBankAccountModel
+            {
+                CurrencyType = currency,
+                Amount = -amountSubstracted,
+                Email = user.Email
+            };
+
+            BankAccount currentAccount = _dbContext.BankAccount.SingleOrDefault(account => account.CurrencyType == updatedAccount.CurrencyType && account.User == user);
+
+            if ( (currentAccount.Balance + updatedAccount.Amount) < 0 )
```
Comment (Slitthe) [inline Server/VikingVault.Services/BankAccountService.cs]: Wrap the if in some curly brackets please, consistency is good

File: Server/VikingVault.Services/UserCardService.cs [hunk 15]
```diff
@@ -0,0 +1,59 @@
...
+                throw new DatabaseException("Server error.");
+            }
+        }
+
+        public int? FindUserIdByCardNumber(string cardNumber)
+        {
+            try
+            {
+                Card card = _dbContext.Cards.SingleOrDefault(c => c.CardNumber == cardNumber);
+                if (card == null)
```
Comment (Slitthe) [inline Server/VikingVault.Services/UserCardService.cs:38]: Same here

File: Server/VikingVault.API/Controllers/TransferRequestsController.cs (hunk 13 above)
Comment (radulambrino) [reply Server/VikingVault.API/Controllers/TransferRequestsController.cs:64]: Fair enough. Changed it

**Pull request merged**